pkg/apiserver/api_installer.go

```diff
@@ -0,0 +1,246 @@
+
+import (
+	"net/http"
+	"reflect"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+
+	"github.com/emicklei/go-restful"
+	"github.com/golang/glog"
+)
+
+type APIInstaller struct {
+	prefix      string // Path prefix where API resources are to be registered.
+	version     string // The API version being installed.
+	restHandler *RESTHandler
+}
+
+// Installs handlers for API resources.
+func (a *APIInstaller) Install() (ws *restful.WebService, errors []error) {
+	errors = make([]error, 0)
+
+	// Create the WebService.
+	ws = a.newWebService()
+
+	// Initialize the custom handlers.
+	watchHandler := (&WatchHandler{
+		storage:         a.restHandler.storage,
+		codec:           a.restHandler.codec,
+		canonicalPrefix: a.restHandler.canonicalPrefix,
+		selfLinker:      a.restHandler.selfLinker,
+	})
+	redirectHandler := (&RedirectHandler{a.restHandler.storage, a.restHandler.codec})
+	proxyHandler := (&ProxyHandler{a.prefix + "/proxy/", a.restHandler.storage, a.restHandler.codec})
+
+	for path, storage := range a.restHandler.storage {
+		// register legacy patterns where namespace is optional in path
+		if err := a.registerResourceHandlers(path, storage, ws, watchHandler, redirectHandler, proxyHandler, false); err != nil {
+			errors = append(errors, err)
+		}
+		// register patterns where namespace is required in path
+		if err := a.registerResourceHandlers(path, storage, ws, watchHandler, redirectHandler, proxyHandler, true); err != nil {
+			errors = append(errors, err)
+		}
+	}
+	return ws, errors
+}
+
+func (a *APIInstaller) newWebService() *restful.WebService {
+	ws := new(restful.WebService)
+	ws.Path(a.prefix)
+	ws.Doc("API at " + a.prefix + " version " + a.version)
+	// TODO: change to restful.MIME_JSON when we set content type in client
+	ws.Consumes("*/*")
+	ws.Produces(restful.MIME_JSON)
+	ws.ApiVersion(a.version)
+	return ws
+}
+
+// Registers the appropriate handlers for the given REST resource.
+func (a *APIInstaller) registerResourceHandlers(path string, storage RESTStorage, ws *restful.WebService, watchHandler http.Handler, redirectHandler http.Handler, proxyHandler http.Handler, namespaceScope bool) error {
+	// Handler for standard REST verbs (GET, PUT, POST, DELETE).
+	restVerbHandler := restfulStripPrefix(a.prefix, a.restHandler)
+	object := storage.New()
+	// TODO: add scheme to APIInstaller rather than using api.Scheme
+	_, kind, err := api.Scheme.ObjectVersionAndKind(object)
+	if err != nil {
+		return err
+	}
+	versionedPtr, err := api.Scheme.New(a.version, kind)
+	if err != nil {
+		return err
+	}
+	versionedObject := indirectArbitraryPointer(versionedPtr)
+
+	if namespaceScope {
+		path = "ns/{namespace}/" + path
+	}
+
+	glog.V(5).Infof("Installing version=/%s, kind=/%s, path=/%s", a.version, kind, path)
+
+	nameParam := ws.PathParameter("name", "name of the "+kind).DataType("string")
+	namespaceParam := ws.PathParameter("namespace", "object name and auth scope, such as for teams and projects").DataType("string")
```

Comment:
ws.PathParameter is not what it seems :)
It does not add that parameter to ws, but just returns a PathParameter object. ws is unchanged.
The parameter is added to the RouteBuilder later on using RouteBuilder.Param()